x-pack/legacy/plugins/maps/public/layers/vector_layer.js

```diff
@@ -484,12 +490,84 @@ export class VectorLayer extends AbstractLayer {
     }
   }
 
+  async _syncSourceFormatters(syncContext) {
+    if (this._style.constructor.type !== LAYER_STYLE_TYPE.VECTOR) {
+      return;
+    }
+
+    return this._syncFormatters({
+      source: this._source,
+      dataRequestId: SOURCE_FORMATTERS_ID_ORIGIN,
+      dynamicStyleProps: this._style.getDynamicPropertiesArray().filter(dynamicStyleProp => {
+        return dynamicStyleProp.getFieldOrigin() === FIELD_ORIGIN.SOURCE;
+      }),
+      ...syncContext,
+    });
+  }
+
+  async _syncJoinFormatters(syncContext, join) {
+    const joinSource = join.getRightJoinSource();
+    return this._syncFormatters({
+      source: joinSource,
+      dataRequestId: join.getSourceFormattersDataRequestId(),
+      dynamicStyleProps: this._style.getDynamicPropertiesArray().filter(dynamicStyleProp => {
+        const matchingField = joinSource.getMetricFieldForName(
+          dynamicStyleProp.getField().getName()
+        );
```

Comment:
This checking on field-existence is pretty common, and is used multiple times to filter out fields. I'd  consider adding a helper `AbstractESAggSource#hasMatchingMetricField(fieldName)` to wrap this code.

- https://github.com/elastic/kibana/blob/1adadcd18b6029a61404ccd4126d4367bec44067/x-pack/legacy/plugins/maps/public/layers/styles/vector/vector_style.js#L315-L316
- https://github.com/elastic/kibana/blob/1adadcd18b6029a61404ccd4126d4367bec44067/x-pack/legacy/plugins/maps/public/layers/styles/vector/vector_style.js#L583-L584

Reply:
This is actually a little bit more complicated then just finding the matching field.

in `_makeField` the entire method looks like the below so you can not just swap out with a function call. I vote to leave this as is for now

```
      let matchingField = null;
      const joins = this._layer.getValidJoins();
      joins.find(join => {
        const aggSource = join.getRightJoinSource();
        matchingField = aggSource.getMetricFieldForName(fieldDescriptor.name);
        return !!matchingField;
      });
      return matchingField;
```